src/unsync/mpsc.rs

```diff
@@ -0,0 +1,198 @@
+//! A multi-producer, single-consumer, futures-aware, FIFO queue with back
+//! pressure, for use communicating between tasks on the same thread.
+
+use std::fmt;
+use std::error::Error;
+use std::any::Any;
+use std::collections::VecDeque;
+use std::rc::{Rc, Weak};
+use std::cell::RefCell;
+
+use task::{self, Task};
+use {Async, AsyncSink, Poll, StartSend, Sink, Stream};
+
+/// Creates a bounded in-memory channel with buffered storage.
+///
+/// This method creates concrete implementations of the `Stream` and `Sink`
+/// traits which can be used to communicate a stream of values between tasks
+/// with backpressure. The channel capacity is exactly `buffer`. On average,
+/// sending a message through this channel performs no dynamic allocation.
+pub fn channel<T>(buffer: usize) -> (Sender<T>, Receiver<T>) { channel_(Some(buffer)) }
+
+fn channel_<T>(buffer: Option<usize>) -> (Sender<T>, Receiver<T>) {
+    let shared = Rc::new(RefCell::new(Shared {
+        buffer: VecDeque::new(),
+        capacity: buffer,
+        blocked_senders: VecDeque::new(),
+        blocked_recv: None,
+    }));
+    let sender = Sender { shared: Rc::downgrade(&shared) };
+    let receiver = Receiver { shared: shared };
+    (sender, receiver)
+}
+
+struct Shared<T> {
+    buffer: VecDeque<T>,
+    capacity: Option<usize>,
+    blocked_senders: VecDeque<Task>,
+    blocked_recv: Option<Task>,
+}
+
+/// The transmission end of a channel.
+///
+/// This is created by the `channel` function.
+pub struct Sender<T> {
+    shared: Weak<RefCell<Shared<T>>>,
+}
+
+impl<T> Drop for Sender<T> {
+    fn drop(&mut self) {
+        if let Some(task) = self.shared.upgrade().and_then(|shared| shared.borrow_mut().blocked_recv.take()) {
+            // Wake up receiver to avoid deadlock in case this was the last sender
+            task.unpark();
```

Comment:
Perhaps a count could be maintained to ensure this only happens when the last sender is dropped?